src/shell.js

```diff
@@ -193,11 +193,34 @@ if (ENVIRONMENT_IS_NODE) {
   if (typeof process == 'undefined' || !process.release || process.release.name !== 'node') throw new Error('not compiled for this environment (did you build to HTML and try to run it not on the web, or set ENVIRONMENT to something - like node - and run it someplace else - like on the web?)');
 #endif
 #endif
+  // `require()` is no-op in an ESM module, use `createRequire()` to construct
+  // the require()` function.  This is only necessary for multi-environment
+  // builds, `-sENVIRONMENT=node` emits a static import declaration instead.
+  // TODO: Swap all `require()`'s with `import()`'s?
+#if EXPORT_ES6 && ENVIRONMENT_MAY_BE_WEB
+  const { createRequire } = await import('module');
+  /** @suppress{duplicate} */
+  var require = createRequire(import.meta.url);
+#endif
+  // These modules will usually be used on Node.js. Load them eagerly to avoid
+  // the complexity of lazy-loading.
+  var fs = require('fs');
+  var nodePath = require('path');
+
   if (ENVIRONMENT_IS_WORKER) {
-    scriptDirectory = require('path').dirname(scriptDirectory) + '/';
+    scriptDirectory = nodePath.dirname(scriptDirectory);
   } else {
-    scriptDirectory = __dirname + '/';
+#if EXPORT_ES6
+    // EXPORT_ES6 + ENVIRONMENT_IS_NODE always requires use of import.meta.url,
+    // since there's no way getting the current absolute path of the module when
+    // support for that is not available.
+    scriptDirectory = nodePath.dirname(import.meta.url);
```

Comment:
`nodePath` might not always work correctly on URLs. I'd rather use `URL` itself for url manipulation, so here it'd be `new URL('./', import.meta.url).href` (which also ensures trailing slash).

Reply:
Ah, nice! Let me incoperate that.

Reply:
We should probably update the web version of this at some point too. https://github.com/kleisauke/emscripten/blob/cff5945db24ffc560421f28759efc2993f931925/src/shell.js#L401

Although maybe it was written that way because `URL` class didn't exist in some older browsers yet... Just something to look into in the future.

Reply:
I mean, in the end, ideally `scriptDirectory` should be always inferred from `import.meta.url` on any `EXPORT_ES6` + (use meta) target. That's what I tried to do first in my attempt at tackling that issue and failed because of many conflicting codepaths 😅

But we can leave that out of scope.

Reply:
Indeed, `_scriptDir` and `scriptDirectory` needs some careful refactoring. I didn't investigate why this was separated into two vars initially, but perhaps I should.

For now, I incorporated your first suggestion with commit 85a9619c414757a6cfbd784fd8a57541527d30d0.